book-content/chapters/14-configuring-typescript.md

```diff
@@ -0,0 +1,1024 @@
+
+```typescript
+interface VinylSingle {
+  title: string;
+  artist: string;
+  tracks: string[];
+}
+
+const egoMirror: VinylSingle = {
+  title: "Ego / Mirror",
+  artist: "Burial / Four Tet / Thom Yorke",
+  tracks: ["Ego", "Mirror"],
+};
+```
+
+To accessing the b-side of `egoMirror`, we would index into its `tracks` like this:
+
+```typescript
+const bSide = egoMirror.tracks[1];
+console.log(bSide.toUpperCase()); // 'MIRROR'
+```
+
+Without `noUncheckedIndexedAccess` enabled in `tsconfig.json`, TypeScript assumes that indexing will always return a valid value, even if the index is out of bounds.
+
+Trying to access a non-existent fourth track would not raise an error in VS Code, but it does result in a runtime error:
+
+```typescript
+const nonExistentTrack = egoMirror.tracks[3];
+console.log(nonExistentTrack.toUpperCase()); // no error in VS Code
+
+// However, running the code results in a runtime error:
+TypeError: Cannot read property 'toUpperCase' of undefined
+```
+
+By setting `noUncheckedIndexedAccess` to `true`, TypeScript will infer the type of every indexed access to be `T | undefined` instead of just `T`. In this case, every entry in `egoMirror.tracks` would be of type `string | undefined`:
+
+```ts twoslash
+// @noUncheckedIndexedAccess: true
+interface VinylSingle {
+  title: string;
+  artist: string;
+  tracks: string[];
+}
+
+const egoMirror: VinylSingle = {
+  title: "Ego / Mirror",
+  artist: "Burial / Four Tet / Thom Yorke",
+  tracks: ["Ego", "Mirror"],
+};
+
+// ---cut---
+const ego = egoMirror.tracks[0];
+//    ^?
+const mirror = egoMirror.tracks[1];
+const nonExistentTrack = egoMirror.tracks[3];
+```
+
+However, because the types of each of the tracks are now `string | undefined`, we have errors when attempting to call `toUpperCase` even for the valid tracks:
+
+```ts twoslash
+// @errors: 18048
+// @noUncheckedIndexedAccess: true
+interface VinylSingle {
+  title: string;
+  artist: string;
+  tracks: string[];
+}
+
+const egoMirror: VinylSingle = {
+  title: "Ego / Mirror",
+  artist: "Burial / Four Tet / Thom Yorke",
+  tracks: ["Ego", "Mirror"],
+};
+
+const ego = egoMirror.tracks[0];
+// ---cut---
+console.log(ego.toUpperCase()); // red squiggly line under ego
+```
+
+This means that we have to handle the possibility of `undefined` values when accessing array or object indices.
+
+So `noUncheckedIndexedAccess` makes TypeScript more strict, but at the cost of having to handle `undefined` values more carefully.
+
+Usually, this is a good trade-off, as it helps catch potential runtime errors early in the development process. But I wouldn't blame you if you end up turning it off in some cases.
+
+### Other Strictness Options
+
+I tend to configure my `tsconfig.json` no stricter than `strict` and `noUncheckedIndexedAccess`. If you want to go further, there are several other strictness options you can enable:
+
+- `allowUnreachableCode`: Errors when unreachable code is detected, like code after a `return` statement.
```

Comment:
this is already the default... so if anything could disable it and not enable it as an extra strict option